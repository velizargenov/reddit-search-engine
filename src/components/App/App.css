.app {
  color: blue;
}
.wrapper {
  position: relative;
}