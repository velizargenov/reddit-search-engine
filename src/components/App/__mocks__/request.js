export default [
  {
    "kind": "t3",
    "data": {
      "approved_at_utc": null,
      "subreddit": "aww",
      "selftext": "",
      "author_fullname": "t2_g5dnz",
      "saved": false,
      "mod_reason_title": null,
      "gilded": 2,
      "clicked": false,
      "title": "I like to put hats on my grandparents dog, and she likes to wear them",
      "link_flair_richtext": [],
      "subreddit_name_prefixed": "r/aww",
      "hidden": false,
      "pwls": 6,
      "link_flair_css_class": null,
      "downs": 0,
      "thumbnail_height": 140,
      "hide_score": false,
      "name": "t3_c1hnhy",
      "quarantine": false,
      "link_flair_text_color": "dark",
      "author_flair_background_color": null,
      "subreddit_type": "public",
      "ups": 92579,
      "total_awards_received": 7,
      "media_embed": {
        "content": "&lt;iframe class=\"embedly-embed\" src=\"https://cdn.embedly.com/widgets/media.html?src=https%3A%2F%2Fimgur.com%2FXvOZn5K%2Fembed%3Fpub%3Dtrue%26ref%3Dhttps%253A%252F%252Fembed.ly%26w%3D900&amp;url=https%3A%2F%2Fimgur.com%2FXvOZn5K&amp;image=https%3A%2F%2Fi.imgur.com%2FXvOZn5K.jpg%3Ffb&amp;key=ed8fa8699ce04833838e66ce79ba05f1&amp;type=text%2Fhtml&amp;schema=imgur\" width=\"600\" height=\"1397\" scrolling=\"no\" frameborder=\"0\" allow=\"autoplay; fullscreen\" allowfullscreen=\"true\"&gt;&lt;/iframe&gt;",
        "width": 600,
        "scrolling": false,
        "height": 1397
      },
      "thumbnail_width": 140,
      "author_flair_template_id": null,
      "is_original_content": false,
      "user_reports": [],
      "secure_media": {
        "oembed": {
          "provider_url": "http://imgur.com",
          "description": "Post with 0 votes and 0 views.",
          "title": "Imgur",
          "url": "https://imgur.com/XvOZn5K",
          "thumbnail_width": 600,
          "height": 1397,
          "width": 600,
          "html": "&lt;iframe class=\"embedly-embed\" src=\"https://cdn.embedly.com/widgets/media.html?src=https%3A%2F%2Fimgur.com%2FXvOZn5K%2Fembed%3Fpub%3Dtrue%26ref%3Dhttps%253A%252F%252Fembed.ly%26w%3D900&amp;url=https%3A%2F%2Fimgur.com%2FXvOZn5K&amp;image=https%3A%2F%2Fi.imgur.com%2FXvOZn5K.jpg%3Ffb&amp;key=ed8fa8699ce04833838e66ce79ba05f1&amp;type=text%2Fhtml&amp;schema=imgur\" width=\"600\" height=\"1397\" scrolling=\"no\" frameborder=\"0\" allow=\"autoplay; fullscreen\" allowfullscreen=\"true\"&gt;&lt;/iframe&gt;",
          "version": "1.0",
          "provider_name": "Imgur",
          "thumbnail_url": "https://i.imgur.com/XvOZn5K.jpg?fb",
          "type": "rich",
          "thumbnail_height": 315
        },
        "type": "imgur.com"
      },
      "is_reddit_media_domain": false,
      "is_meta": false,
      "category": null,
      "secure_media_embed": {
        "content": "&lt;iframe class=\"embedly-embed\" src=\"https://cdn.embedly.com/widgets/media.html?src=https%3A%2F%2Fimgur.com%2FXvOZn5K%2Fembed%3Fpub%3Dtrue%26ref%3Dhttps%253A%252F%252Fembed.ly%26w%3D900&amp;url=https%3A%2F%2Fimgur.com%2FXvOZn5K&amp;image=https%3A%2F%2Fi.imgur.com%2FXvOZn5K.jpg%3Ffb&amp;key=ed8fa8699ce04833838e66ce79ba05f1&amp;type=text%2Fhtml&amp;schema=imgur\" width=\"600\" height=\"1397\" scrolling=\"no\" frameborder=\"0\" allow=\"autoplay; fullscreen\" allowfullscreen=\"true\"&gt;&lt;/iframe&gt;",
        "width": 600,
        "scrolling": false,
        "media_domain_url": "https://www.redditmedia.com/mediaembed/c1hnhy",
        "height": 1397
      },
      "link_flair_text": null,
      "can_mod_post": false,
      "score": 92579,
      "approved_by": null,
      "thumbnail": "https://b.thumbs.redditmedia.com/cCDPS-jcyXj8Z26rtQmG3kGdB3ykjRNlO0NWKjVpf5Q.jpg",
      "edited": false,
      "author_flair_css_class": null,
      "author_flair_richtext": [],
      "gildings": {
        "gid_1": 5,
        "gid_2": 2
      },
      "post_hint": "link",
      "content_categories": null,
      "is_self": false,
      "mod_note": null,
      "created": 1560763898,
      "link_flair_type": "text",
      "wls": 6,
      "banned_by": null,
      "author_flair_type": "text",
      "domain": "imgur.com",
      "selftext_html": null,
      "likes": null,
      "suggested_sort": null,
      "banned_at_utc": null,
      "view_count": null,
      "archived": false,
      "no_follow": false,
      "is_crosspostable": false,
      "pinned": false,
      "over_18": false,
      "preview": {
        "images": [
          {
            "source": {
              "url": "https://external-preview.redd.it/s5YhdI1oeNxU_66essDrss6jRYlOwdulqXJxpdcb_iM.jpg?auto=webp&amp;s=8f7645fb3182fc9b5f3440d099af89be09909bde",
              "width": 1717,
              "height": 3629
            },
            "resolutions": [
              {
                "url": "https://external-preview.redd.it/s5YhdI1oeNxU_66essDrss6jRYlOwdulqXJxpdcb_iM.jpg?width=108&amp;crop=smart&amp;auto=webp&amp;s=cacd40a7c5b28c678c567a63b96794b7efb3fbdc",
                "width": 108,
                "height": 216
              },
              {
                "url": "https://external-preview.redd.it/s5YhdI1oeNxU_66essDrss6jRYlOwdulqXJxpdcb_iM.jpg?width=216&amp;crop=smart&amp;auto=webp&amp;s=536b22d1a9fe07ab6b6828129d898aefd3c360d4",
                "width": 216,
                "height": 432
              },
              {
                "url": "https://external-preview.redd.it/s5YhdI1oeNxU_66essDrss6jRYlOwdulqXJxpdcb_iM.jpg?width=320&amp;crop=smart&amp;auto=webp&amp;s=9f9dfc829963fbe4479fb1b8540d080828992fd6",
                "width": 320,
                "height": 640
              },
              {
                "url": "https://external-preview.redd.it/s5YhdI1oeNxU_66essDrss6jRYlOwdulqXJxpdcb_iM.jpg?width=640&amp;crop=smart&amp;auto=webp&amp;s=0c9c3fdaf2b23eecc015a4cd8fbbef85600bacde",
                "width": 640,
                "height": 1280
              },
              {
                "url": "https://external-preview.redd.it/s5YhdI1oeNxU_66essDrss6jRYlOwdulqXJxpdcb_iM.jpg?width=960&amp;crop=smart&amp;auto=webp&amp;s=169625df52fa747dc4a636e3c4fbd44add3a41c1",
                "width": 960,
                "height": 1920
              },
              {
                "url": "https://external-preview.redd.it/s5YhdI1oeNxU_66essDrss6jRYlOwdulqXJxpdcb_iM.jpg?width=1080&amp;crop=smart&amp;auto=webp&amp;s=45dcbe350a94d5eefe8caa853632ae12c8cb4d5a",
                "width": 1080,
                "height": 2160
              }
            ],
            "variants": {},
            "id": "QWeGRte7f8nEmPWPoXLN3xlV6HFHmqzGV-JI70nd5-Q"
          }
        ],
        "enabled": false
      },
      "all_awardings": [
        {
          "is_enabled": true,
          "count": 2,
          "subreddit_id": null,
          "description": "Gold Award",
          "coin_reward": 100,
          "icon_width": 512,
          "icon_url": "https://www.redditstatic.com/gold/awards/icon/gold_512.png",
          "days_of_premium": 7,
          "icon_height": 512,
          "resized_icons": [
            {
              "url": "https://www.redditstatic.com/gold/awards/icon/gold_16.png",
              "width": 16,
              "height": 16
            },
            {
              "url": "https://www.redditstatic.com/gold/awards/icon/gold_32.png",
              "width": 32,
              "height": 32
            },
            {
              "url": "https://www.redditstatic.com/gold/awards/icon/gold_48.png",
              "width": 48,
              "height": 48
            },
            {
              "url": "https://www.redditstatic.com/gold/awards/icon/gold_64.png",
              "width": 64,
              "height": 64
            },
            {
              "url": "https://www.redditstatic.com/gold/awards/icon/gold_128.png",
              "width": 128,
              "height": 128
            }
          ],
          "days_of_drip_extension": 0,
          "award_type": "global",
          "coin_price": 500,
          "id": "gid_2",
          "name": "Gold"
        },
        {
          "is_enabled": true,
          "count": 5,
          "subreddit_id": null,
          "description": "Silver Award",
          "coin_reward": 0,
          "icon_width": 512,
          "icon_url": "https://www.redditstatic.com/gold/awards/icon/silver_512.png",
          "days_of_premium": 0,
          "icon_height": 512,
          "resized_icons": [
            {
              "url": "https://www.redditstatic.com/gold/awards/icon/silver_16.png",
              "width": 16,
              "height": 16
            },
            {
              "url": "https://www.redditstatic.com/gold/awards/icon/silver_32.png",
              "width": 32,
              "height": 32
            },
            {
              "url": "https://www.redditstatic.com/gold/awards/icon/silver_48.png",
              "width": 48,
              "height": 48
            },
            {
              "url": "https://www.redditstatic.com/gold/awards/icon/silver_64.png",
              "width": 64,
              "height": 64
            },
            {
              "url": "https://www.redditstatic.com/gold/awards/icon/silver_128.png",
              "width": 128,
              "height": 128
            }
          ],
          "days_of_drip_extension": 0,
          "award_type": "global",
          "coin_price": 100,
          "id": "gid_1",
          "name": "Silver"
        }
      ],
      "media_only": false,
      "can_gild": false,
      "spoiler": false,
      "locked": false,
      "author_flair_text": null,
      "visited": false,
      "num_reports": null,
      "distinguished": null,
      "subreddit_id": "t5_2qh1o",
      "mod_reason_by": null,
      "removal_reason": null,
      "link_flair_background_color": "",
      "id": "c1hnhy",
      "is_robot_indexable": true,
      "report_reasons": null,
      "author": "FuckinMagic",
      "num_crossposts": 29,
      "num_comments": 768,
      "send_replies": true,
      "whitelist_status": "all_ads",
      "contest_mode": false,
      "mod_reports": [],
      "author_patreon_flair": false,
      "author_flair_text_color": null,
      "permalink": "/r/aww/comments/c1hnhy/i_like_to_put_hats_on_my_grandparents_dog_and_she/",
      "parent_whitelist_status": "all_ads",
      "stickied": false,
      "url": "https://imgur.com/XvOZn5K",
      "subreddit_subscribers": 20918619,
      "created_utc": 1560735098,
      "media": {
        "oembed": {
          "provider_url": "http://imgur.com",
          "description": "Post with 0 votes and 0 views.",
          "title": "Imgur",
          "url": "https://imgur.com/XvOZn5K",
          "thumbnail_width": 600,
          "height": 1397,
          "width": 600,
          "html": "&lt;iframe class=\"embedly-embed\" src=\"https://cdn.embedly.com/widgets/media.html?src=https%3A%2F%2Fimgur.com%2FXvOZn5K%2Fembed%3Fpub%3Dtrue%26ref%3Dhttps%253A%252F%252Fembed.ly%26w%3D900&amp;url=https%3A%2F%2Fimgur.com%2FXvOZn5K&amp;image=https%3A%2F%2Fi.imgur.com%2FXvOZn5K.jpg%3Ffb&amp;key=ed8fa8699ce04833838e66ce79ba05f1&amp;type=text%2Fhtml&amp;schema=imgur\" width=\"600\" height=\"1397\" scrolling=\"no\" frameborder=\"0\" allow=\"autoplay; fullscreen\" allowfullscreen=\"true\"&gt;&lt;/iframe&gt;",
          "version": "1.0",
          "provider_name": "Imgur",
          "thumbnail_url": "https://i.imgur.com/XvOZn5K.jpg?fb",
          "type": "rich",
          "thumbnail_height": 315
        },
        "type": "imgur.com"
      },
      "is_video": false
    }
  },
  {
    "kind": "t3",
    "data": {
      "approved_at_utc": null,
      "subreddit": "aww",
      "selftext": "",
      "author_fullname": "t2_j7bq4",
      "saved": false,
      "mod_reason_title": null,
      "gilded": 0,
      "clicked": false,
      "title": "My foster kitten, Gabe, had a rough day today.",
      "link_flair_richtext": [],
      "subreddit_name_prefixed": "r/aww",
      "hidden": false,
      "pwls": 6,
      "link_flair_css_class": null,
      "downs": 0,
      "thumbnail_height": 78,
      "hide_score": false,
      "name": "t3_c1h7hv",
      "quarantine": false,
      "link_flair_text_color": "dark",
      "author_flair_background_color": null,
      "subreddit_type": "public",
      "ups": 11714,
      "total_awards_received": 0,
      "media_embed": {},
      "thumbnail_width": 140,
      "author_flair_template_id": null,
      "is_original_content": false,
      "user_reports": [],
      "secure_media": {
        "reddit_video": {
          "fallback_url": "https://v.redd.it/7d3h4s0pct431/DASH_1080?source=fallback",
          "height": 1080,
          "width": 1920,
          "scrubber_media_url": "https://v.redd.it/7d3h4s0pct431/DASH_96",
          "dash_url": "https://v.redd.it/7d3h4s0pct431/DASHPlaylist.mpd",
          "duration": 26,
          "hls_url": "https://v.redd.it/7d3h4s0pct431/HLSPlaylist.m3u8",
          "is_gif": true,
          "transcoding_status": "completed"
        }
      },
      "is_reddit_media_domain": true,
      "is_meta": false,
      "category": null,
      "secure_media_embed": {},
      "link_flair_text": null,
      "can_mod_post": false,
      "score": 11714,
      "approved_by": null,
      "thumbnail": "https://a.thumbs.redditmedia.com/LusQn5Z0vYTsMWBQRCk4LW-PjItYAFRouiOxJCnrwU0.jpg",
      "edited": false,
      "author_flair_css_class": null,
      "author_flair_richtext": [],
      "gildings": {},
      "post_hint": "hosted:video",
      "content_categories": null,
      "is_self": false,
      "mod_note": null,
      "created": 1560761125,
      "link_flair_type": "text",
      "wls": 6,
      "banned_by": null,
      "author_flair_type": "text",
      "domain": "v.redd.it",
      "selftext_html": null,
      "likes": null,
      "suggested_sort": null,
      "banned_at_utc": null,
      "view_count": null,
      "archived": false,
      "no_follow": false,
      "is_crosspostable": false,
      "pinned": false,
      "over_18": false,
      "preview": {
        "images": [
          {
            "source": {
              "url": "https://external-preview.redd.it/ZtAdZpzVDFuviL56tkMZBx-OksVU2lYjv5e6HqwYm08.png?format=pjpg&amp;auto=webp&amp;s=99096ea67b203c46f6ac1dc9738ca5fb43853dc7",
              "width": 1920,
              "height": 1080
            },
            "resolutions": [
              {
                "url": "https://external-preview.redd.it/ZtAdZpzVDFuviL56tkMZBx-OksVU2lYjv5e6HqwYm08.png?width=108&amp;crop=smart&amp;format=pjpg&amp;auto=webp&amp;s=4fc6a2daf5c349b1955908e22d59ed760ad9783a",
                "width": 108,
                "height": 60
              },
              {
                "url": "https://external-preview.redd.it/ZtAdZpzVDFuviL56tkMZBx-OksVU2lYjv5e6HqwYm08.png?width=216&amp;crop=smart&amp;format=pjpg&amp;auto=webp&amp;s=2bbf96bf13491737b814c17fa62e05ce880e66ad",
                "width": 216,
                "height": 121
              },
              {
                "url": "https://external-preview.redd.it/ZtAdZpzVDFuviL56tkMZBx-OksVU2lYjv5e6HqwYm08.png?width=320&amp;crop=smart&amp;format=pjpg&amp;auto=webp&amp;s=d3fc43357359471b9e75a33be06c600f4a9fd9b6",
                "width": 320,
                "height": 180
              },
              {
                "url": "https://external-preview.redd.it/ZtAdZpzVDFuviL56tkMZBx-OksVU2lYjv5e6HqwYm08.png?width=640&amp;crop=smart&amp;format=pjpg&amp;auto=webp&amp;s=64803457d1405216ea67e67819c05d9cfba6afb8",
                "width": 640,
                "height": 360
              },
              {
                "url": "https://external-preview.redd.it/ZtAdZpzVDFuviL56tkMZBx-OksVU2lYjv5e6HqwYm08.png?width=960&amp;crop=smart&amp;format=pjpg&amp;auto=webp&amp;s=851ecd6fe72f76f6c68cc3974937a8e4498afe3c",
                "width": 960,
                "height": 540
              },
              {
                "url": "https://external-preview.redd.it/ZtAdZpzVDFuviL56tkMZBx-OksVU2lYjv5e6HqwYm08.png?width=1080&amp;crop=smart&amp;format=pjpg&amp;auto=webp&amp;s=b1430a75f007f9cd4b21e447cc400db93b9485d9",
                "width": 1080,
                "height": 607
              }
            ],
            "variants": {},
            "id": "fYp87SAhL7O95ftbvmw1qM-grU32j6m84OYs4ptF0qo"
          }
        ],
        "enabled": false
      },
      "all_awardings": [],
      "media_only": false,
      "can_gild": false,
      "spoiler": false,
      "locked": false,
      "author_flair_text": null,
      "visited": false,
      "num_reports": null,
      "distinguished": null,
      "subreddit_id": "t5_2qh1o",
      "mod_reason_by": null,
      "removal_reason": null,
      "link_flair_background_color": "",
      "id": "c1h7hv",
      "is_robot_indexable": true,
      "report_reasons": null,
      "author": "ldenzel",
      "num_crossposts": 2,
      "num_comments": 79,
      "send_replies": true,
      "whitelist_status": "all_ads",
      "contest_mode": false,
      "mod_reports": [],
      "author_patreon_flair": false,
      "author_flair_text_color": null,
      "permalink": "/r/aww/comments/c1h7hv/my_foster_kitten_gabe_had_a_rough_day_today/",
      "parent_whitelist_status": "all_ads",
      "stickied": false,
      "url": "https://v.redd.it/7d3h4s0pct431",
      "subreddit_subscribers": 20918619,
      "created_utc": 1560732325,
      "media": {
        "reddit_video": {
          "fallback_url": "https://v.redd.it/7d3h4s0pct431/DASH_1080?source=fallback",
          "height": 1080,
          "width": 1920,
          "scrubber_media_url": "https://v.redd.it/7d3h4s0pct431/DASH_96",
          "dash_url": "https://v.redd.it/7d3h4s0pct431/DASHPlaylist.mpd",
          "duration": 26,
          "hls_url": "https://v.redd.it/7d3h4s0pct431/HLSPlaylist.m3u8",
          "is_gif": true,
          "transcoding_status": "completed"
        }
      },
      "is_video": true
    }
  }
]