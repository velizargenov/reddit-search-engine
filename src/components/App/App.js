import React, { Component, Fragment  } from 'react';

import { css } from '@emotion/core';
import { ClipLoader } from 'react-spinners';

import './App.css';
import Search from '../Search/Search';
import PostsList from '../PostsList/PostsList';

const override = css`
  position: absolute;
  top: 100px;
  margin-left: 50%;
  display: block;
  margin: 0 auto;
  left: 50%;
`;

class App extends Component {
  constructor(props) {
    super(props);

    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleChange = this.handleChange.bind(this);

    this.state = {
      value: '',
      data: [],
      loading: false
    }
  }

  handleChange (event) {
    this.setState({
      value: event.target.value
    });
  }

  handleSubmit (event) {
    event.preventDefault();
    const { value } = this.state;
    this.setState({
      data: [],
      loading: true
    })

    if (value !== '') {
      this.retrieveData(value)
        .then(data => this.setState({ data, loading: false }));
    }
  }

  async retrieveData (searchValue) {
    const response = await fetch(`https://www.reddit.com/r/${searchValue}.json`);
    const data = await response.json();

    return data && data.data && data.data.children;
  }

  render () {
    return (
      <div className="wrapper">
        <Search
          value={this.state.value}
          handleSubmit={this.handleSubmit}
          handleChange={this.handleChange}
        />
        <PostsList posts={this.state.data} />
        <ClipLoader
          css={override}
          sizeUnit={"px"}
          size={150}
          color={'#123abc'}
          loading={this.state.loading}
        />
      </div>
    )
  }
}

export default App;
