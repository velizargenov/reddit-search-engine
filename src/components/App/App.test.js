import React from 'react';
import toJson from 'enzyme-to-json';
import { shallow, mount } from 'enzyme';

import App from './App';
import request from './__mocks__/request';

function mockFetch(data) {
  return jest.fn().mockImplementation(() =>
    Promise.resolve({
      ok: true,
      json: () => data
    })
  );
}

describe('about App', () => {
  let app;

  beforeEach(() => {
    app = shallow(<App />)
  });

  it('should render component without crashing', () => {
    expect(app.length).toBe(1);
  });

  it('matches snapshot', () => {
    expect(toJson(app)).toMatchSnapshot();
  });

  it("should make one request to the endpoint", async () => {
    fetch = mockFetch(request);
    const appComponent = mount(<App />);

    await appComponent.instance().retrieveData('aww');
    expect(fetch).toHaveBeenCalledTimes(1);
  });
});
