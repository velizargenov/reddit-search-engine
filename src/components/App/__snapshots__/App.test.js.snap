// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`about App matches snapshot 1`] = `
<Fragment>
  <Search
    handleChange={[Function]}
    handleSubmit={[Function]}
    value=""
  />
  <PostsList />
</Fragment>
`;
